File: api-node/src/core/utils/error.util.ts
import { logger } from './logger.util.js';

export class AppError extends Error {
	constructor(
		public message: string,
		public statusCode: number,
		public isOperational = true,
	) {
		super(message);
		Object.setPrototypeOf(this, new.target.prototype);
		Error.captureStackTrace(this);
	}
}

// HTTP Errors
export class BadRequestError extends AppError {
	constructor(message = 'Bad Request') {
		super(message, 400);
	}
}
export class NotAuthorizedError extends AppError {
	constructor(message = 'Not Authorized') {
		super(message, 401);
	}
}
export class ForbiddenError extends AppError {
	constructor(message = 'Forbidden Access') {
		super(message, 403);
	}
}
export class NotFoundError extends AppError {
	constructor(message = 'Not found') {
		super(message, 404);
	}
}
export class ConflictError extends AppError {
	constructor(message = 'Conflict') {
		super(message, 409);
	}
}

// Error handlers
export const repositoryError = (error: any, message: string, fallback: string, context: any): never => {
	logger.error({
		message: 'Repository Layer: ' + message,
		...context,
		error: error.message,
		stack: error.stack,
	});

	if (error instanceof AppError) {
		throw error;
	}

	throw new AppError(fallback, 500);
};

export const serviceError = (error: any, message: string, fallback: string, context: any): never => {
	if (!(error instanceof AppError)) {
		logger.error({
			message,
			...context,
			error: error.message,
			stack: error.stack,
		});
	}

	if (error instanceof AppError) throw error;
	throw new AppError(fallback, 500);
};

File: api-dotnet/Utils/ErrorUtil.cs
public class AppError : Exception
{
    public int StatusCode { get; }

    public AppError(string message, int statusCode, bool isOperational = true) : base(message)
    {
        StatusCode = statusCode;
        Data["isOperational"] = isOperational;
    }
}

public class BadRequestError : AppError
{
    public BadRequestError(string message = "Bad Request") : base(message, 400)
    {
    }
}

public class NotAuthorizedError : AppError
{
    public NotAuthorizedError(string message = "Not Authorized") : base(message, 401)
    {
    }
}

public class ForbiddenError : AppError
{
    public ForbiddenError(string message = "Forbidden Access") : base(message, 403)
    {
    }
}

public class NotFoundError : AppError
{
    public NotFoundError(string message = "Not Found") : base(message, 404)
    {
    }
}

public class ConflictError : AppError
{
    public ConflictError(string message = "Conflict") : base(message, 409)
    {
    }
}
